### PR TITLE
Encryption proposal

### PR DESCRIPTION
**Pull request by aditya-dixit99**
# Proposed Changes

-Proposal for storage encryption

### CODE REVIEW

**Review by afritzler — CHANGES_REQUESTED** [review]
Can we please move this proposal to the `proposals` folder as there might be more coming in the future.

**Comment by pradumnapandit** [comment]
Done, Changed proposal to Proposals.

**Review by schucly — APPROVED** [review]
Looks good for me.

File: docs/proposals/01-storage-encryption.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+- Object storage encryption
+
+### Details
+As of now two types of encryption is supported by Ceph: 
+- OSD Level 
+- Block device Level
+Currently we are moving ahead in this proposal with Block device level.
+  
+## Proposal
+- Data encryption key(DEK) will be provided by user in volume object. Cephlet will internally create a Key encryption key(KEK), encrypt the data encryption key and add     into the metadata. All the changes will take place in cephCSI.
```
Comment (lukasfrank) [inline docs/proposals/01-storage-encryption.md:63]: Did I get it correct that the `Secret` referenced by the `Volume` contains the data encryption key?

Comment (lukasfrank) [inline docs/proposals/01-storage-encryption.md:63]: Where will be the KEK being stored?

Comment (pradumnapandit) [reply docs/proposals/01-storage-encryption.md:63]: Yes.

Comment (pradumnapandit) [reply docs/proposals/01-storage-encryption.md:63]: KEK will be stored as Secret by cephlet.

File: docs/proposals/01-storage-encryption.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+```
+
+Additionally Update the rook-ceph-operator-config configmap and patch the following configurations
+```
+kubectl patch cm rook-ceph-operator-config -nrook-ceph -p $'data:\n "CSI_ENABLE_ENCRYPTION": "true"'
+```
+
+
+```
+apiVersion: v1
```
Comment (hardikdr) [inline docs/proposals/01-storage-encryption.md:120]: Maybe examples could be expanded to also include how the DEK and KEK are stored.

File: docs/proposals/01-storage-encryption.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+- When writing data to the encrypted RBD image, it is automatically encrypted by LUKS before being written to the RBD image. When reading data from the RBD image, it       is decrypted by LUKS on the client system before being made available to the application.
+
+### Way 2:
+### Details
+ As of now two types of encryption is supported by Ceph: 
+- OSD Level and Block device Level
+(Currently we are moving ahead in this proposal with Block device level.)
+
+- Data encryption key(DEK) will be provided by user in volume object. Cephlet will internally create a Key encryption key(KEK), encrypt the data encryption key and add     into the metadata. All the changes will take place in cephCSI.
```
[the remark this answers is not in the document]
Comment (pradumnapandit) [reply docs/proposals/01-storage-encryption.md:89]: Right, User will provide DEK, and Passphrase is always created internally. And Passphrase is the KEK. 
We will clarify more around the KEK point.

File: docs/proposals/01-storage-encryption.md (hunk 2 above)
Comment (DivyaD211093) [reply docs/proposals/01-storage-encryption.md:150]: We have expanded the examples to include volume object which has secret reference to DEK. 
KEK is generated by cephlet and example we have added in proposal. 

FYI, currently the proposal is in revision state and we might need to restructure it according to final decisions.

File: docs/proposals/01-storage-encryption.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+
+## Motivation
+- Security: Security is an important concern and should be a strong focus of any Cloud Native IaaS. To protect the data against variety of security threats, including        data breaches, hacking and physical theft, encryption plays vital role.
+- Privacy: Encryption helps to maintain privacy of data from being accessed by unauthorised parties.
+- Compliance and business cotinuity: In order to meet compliance requirements of businesses keeping the standards and regulations in mind, it becomes essential to          store the encryption enabled data in the cloud so that in the events of disaster and data theft, organizations can recover the data quickly.
```
[the remark this answers is not in the document]
Comment (aditya-dixit99) [reply docs/proposals/01-storage-encryption.md]: Done

File: docs/proposals/01-storage-encryption.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+Overall, the motivation for a storage encryption proposal is to provide a holostic and secure approach to protect sensitive data and ensure compliance with regulations and standards.
```
[the remark this answers is not in the document]
Comment (pradumnapandit) [reply docs/proposals/01-storage-encryption.md]: Done

**Comment by lukasfrank** [comment]
Let's put this PR on hold since it's gonna be implemented in #187

**Comment by pradumnapandit** [comment]
> Let's put this PR on hold since it's gonna be implemented in #187
Noted.

**Comment by lukasfrank** [comment]
Closed since `Volume Encryption` was implemented in PR #245

**Pull request closed without merging**